honeybee_energy/construction.py

```diff
@@ -1,33 +1,1111 @@
+            elif isinstance(mat, EnergyWindowMaterialGlazing):
+                trans *= mat.visible_transmittance
+                diffusing = True if mat.solar_diffusing is True else False
+            elif isinstance(mat, EnergyWindowMaterialShade):
+                trans *= mat.visible_transmittance
+                diffusing = True
+            elif isinstance(mat, EnergyWindowMaterialBlind):
+                raise NotImplementedError('to_radiance_visible() is not supported for '
+                                          'window constructions with blind materials.')
+        if diffusing is False:
+            return Glass.from_single_transmittance(self.name, trans)
+        else:
+            try:
+                ref = self.materials[-1].solar_reflectance_back
+            except AttributeError:
+                ref = self.materials[-1].solar_reflectance
+            return Trans.from_single_reflectance(self.name, rgb_reflectance=ref,
+                                                 transmitted_diff=trans,
+                                                 transmitted_spec=0)
+
+    def to_dict(self):
+        """Window construction dictionary representation."""
+        return {
+            'type': 'EnergyConstructionWindow',
+            'name': self.name,
+            'materials': [m.to_dict() for m in self.materials]
+        }
+
+    @staticmethod
+    def extract_all_from_idf_file(idf_file):
+        """Get all WindowConstruction objects in an EnergyPlus IDF file.
+
+        Args:
+            idf_file: A path to an IDF file containing objects for window
+                constructions and corresponding materials. For example, the
+                IDF Report output by LBNL WINDOW.
+        """
+        # check the file
+        assert os.path.isfile(idf_file), 'Cannot find an idf file at {}'.format(idf_file)
+        with open(idf_file, 'r') as ep_file:
+            file_contents = ep_file.read()
+        # extract all material objects
+        mat_pattern = re.compile(r"(?i)(WindowMaterial:[\s\S]*?;)")
+        material_str = mat_pattern.findall(file_contents)
+        materials_dict = WindowConstruction._idf_materials_dictionary(material_str)
+        materials = list(materials_dict.values())
+        # extract all of the construction objects
+        constr_pattern = re.compile(r"(?i)(Construction,[\s\S]*?;)")
+        constr_props = tuple(WindowConstruction._parse_ep_string(ep_string) for
+                             ep_string in constr_pattern.findall(file_contents))
+        constructions = []
+        for constr in constr_props:
+            try:
+                constr_mats = [materials_dict[mat] for mat in constr[1:]]
+                constructions.append(WindowConstruction(constr[0], constr_mats))
+            except KeyError:
+                pass  # the construction is an opaque construction
+        return constructions, materials
+
+    @staticmethod
+    def _idf_materials_dictionary(ep_mat_strings):
```

Comment:
Ok. You mean that we should have something like [what you have done here](https://github.com/ladybug-tools/honeybee-radiance/blob/master/honeybee_radiance/parser.py)?  We can eventually do that but `honeybee_energy` is different in that not all objects will be backward translatable from IDF.  All of the complex HVAC stuff will never be translatable from IDF but there is also the fact that, given that we are using schedule ruleset objects for most schedules (which exist in OpenStudio but not E+), writing a parser to generate schedule rulesets from the IDF is going to be tough.  The point that I am making is that I am not sure that we will use our own built-in parser for importing anything other than materials/constructions and geometry from IDF files.  Instead, we will probably rely on OpenStudio to translate from IDF or gbXML and then use the OpenStudio model to build the honeybee model.

Reply:
Yes. I was talking about having them more accessible as they will be useful for others.

> All of the complex HVAC stuff will never be translatable from IDF but there is also the schedule ruleset objects that we are using for most schedules (which exist in OpenStudio but not E+). This means that writing a parser to generate schedule rulesets from the IDF is going to be tough and the overall point that I am making is that I am not sure that we will use our own built-in IDF parser for importing anything other than materials/constructions and geometry from IDF files. Instead, we will probably rely on OpenStudio to translate from IDF or gbXML and then use the OpenStudio model to build the honeybee model.

Agreed! That is why I wrote "idf parser for **geometries and materials**" in my comment.

Also I'm not suggesting us writing our own parser, that's why I said "in case **one** wants to write an idf parser for geometries and materials at some point it can be accessed directly.".

I just think this is something which will be useful to use to other users and is better to be exposed better through the API.